projects/construction/projspec_repo2singularity.md

```diff
@@ -0,0 +1,71 @@
+# Project: Repo2singularity
+
+Repo2singularity
+
+## Campaigns:
+
+- WORKFLOWS: Publication
+
+## Overview
+
+[BinderHub](https://github.com/jupyterhub/binderhub) allows users to interact with code and environment from a Git Repo within a JupyterHub instance. To achieve this, binderhub ties together
+
+- [JupyterHub](https://github.com/jupyterhub/jupyterhub): Provides authentication, and spawns single user Jupyter Notebook servers.
+- [Repo2Docker](https://github.com/jupyter/repo2docker): Generates a Docker image from the Git Repo.
+  
+Currently, BinderHub as a Python package can not run on traditional HPC systems. As [Joe Hamman](https://github.com/jhamman) pointed out in [BinderHub for HPC](https://discourse.jupyter.org/t/binderhub-for-hpc/143/4) post on discourse, some of the reasons why it's difficult to run BinderHub on HPC include:
+
+- HPC systems are rarely container friendly (BinderHub requires both Kubernetes and Docker).
+- Often compute nodes do not have access to the outside network.
+- Managing a server attached to an HPC is not going to be popular from sys admins.
+- Provisioning resources requires waiting in a job queue.
+
+Despite these challenges, there's hope that we can re-purpose some key pieces of binderhub, and be able to get *binderhub like functionality on HPC*. [Yuvi Panda](https://github.com/yuvipanda) suggested breaking BinderHub down into two components:
+
+1. Dynamic image building from a git repository. One way to accomplish this on an HPC system is to swap Docker with some HPC compatible container/image builder. For example [Singularity](https://github.com/hpcng/singularity), [Shifter](https://github.com/NERSC/shifter), [Charliecloud](https://github.com/hpc/charliecloud).
+2. Launching an interactive web application from inside the image. This involves some work on JupyterHub's side.  
+
+This project is going to address the first point (1): **Dynamic image building from a git repository**. Since singularity containers are currently supported on Cheyenne (`module load singularity/3.3.0`), this project will implement **repo2singularity**, a sister project of repo2docker that will allow users to turn git repos into singularity images that can be run on Cheyenne, Casper or other HPC systems that support singularity containers.
+
+## Repositories
+
+- https://github.com/jupyter/repo2docker
+- https://github.com/andersy005/repo2singularity
+
+## Skills & Knowledge
+
+- Knowledge of Docker (Required)
+- Knowledge of [singularity](https://github.com/hpcng/singularity) (Required)
+- Knowlegde of [repo2docker](https://github.com/jupyter/repo2docker) (Optional)
+
+## Deliverables
+
+1. Implement `repo2singularity`, a command line tool that will
+
+   - build singularity images from a git/GitHub repo.
+   - push built images to a singularity image registry such as https://cloud.sylabs.io/library
```

Comment:
My understanding is that `repo2docker` does not current run as a "service," correct?  That is, `repo2docker` is used *within* the BinderHub service, but `repo2docker` is not a service itself.

But `repo2singularity` *will* be a service.  Or, at least, it will be tied to a service.  ...?  This is where I'm confused.

It would be nice in this spec if we could show a diagram of what happens when you click a "Binder button" in the `repo2docker`-model vs the `repo2singularity` model.  What is different?  What is the same?

Reply:
See my next comment before responding...

Reply:
> But repo2singularity will be a service. Or, at least, it will be tied to a service. ...? This is where I'm confused.


`repo2singularity` won't be a service. It's just going to be a wrapper around `repo2docker` with one extra step: converting the docker image into Singularity Image File (SIF):

`repo2docker`: 

- Repo (input) ==> [repo2docker] ==> **docker image**(output)

`repo2singularity`: 

- Repo (input) ==> [repo2docker] ==> docker image (intermediate output) ==> [singularity] ==> **singularity image**(output)


In both cases, the final product is just a container/image. 
We will likely need a `jupyterhub-singularityspawner` to enable JupyterHub to spawn single user notebook servers in Singularity containers.

> It would be nice in this spec if we could show a diagram of what happens when you click a "Binder button" in the repo2docker-model vs the repo2singularity model. What is different? What is the same?

I will look into this.